src/pal/pal_apple.h

```diff
@@ -1,92 +1,218 @@
+      int(VM_MAKE_TAG(uint32_t(PALAnonID)));
+
+    // Encoded memory tag passed to `mach_vm_map`.
+    static constexpr int default_mach_vm_map_flags =
+      int(VM_MAKE_TAG(uint32_t(PALAnonID)));
 
-#  if defined(PLATFORM_IS_ARM)
     /**
-     * Overriding here to mark the page as reusable
-     * rolling it as much as necessary.
-     * As above, the x86 h/w worked alright without this change
-     * however now large allocations work better and more reliably
-     * with on ARM, not to mention better RSS number accuracy
-     * for tools based on task_info API.
+     * Notify platform that we will not be using these pages.
      */
     static void notify_not_using(void* p, size_t size) noexcept
     {
-      SNMALLOC_ASSERT(is_aligned_block<PALBSD::page_size>(p, size));
-#    ifdef USE_POSIX_COMMIT_CHECKS
+      SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
       memset(p, 0x5a, size);
-#    endif
+#  endif
+
+      // `MADV_FREE_REUSABLE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      //
+      // macOS 11 Big Sur may behave in an undocumented manner.
       while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
         ;
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // This must occur after `MADV_FREE_REUSABLE`.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_NONE);
+#  endif
     }
 
     /**
-     * same remark as above but we need to mark the page as REUSE
-     * first
+     * Notify platform that we will be using these pages.
      */
     template<ZeroMem zero_mem>
     static void notify_using(void* p, size_t size) noexcept
     {
       SNMALLOC_ASSERT(
-        is_aligned_block<PALBSD::page_size>(p, size) || (zero_mem == NoZero));
+        is_aligned_block<page_size>(p, size) || (zero_mem == NoZero));
+
+      if constexpr (zero_mem == YesZero)
+      {
+        void* r = mmap(
+          p,
+          size,
+          PROT_READ | PROT_WRITE,
+          MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
+          anonymous_memory_fd,
+          0);
+
+        if (likely(r != MAP_FAILED))
+        {
+          return;
+        }
+      }
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // Mark pages as writable for `madvise` below.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_READ | PROT_WRITE);
+#  endif
+
+      // `MADV_FREE_REUSE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      //
+      // macOS 11 Big Sur may behave in an undocumented manner.
       while (madvise(p, size, MADV_FREE_REUSE) == -1 && errno == EAGAIN)
         ;
 
       if constexpr (zero_mem == YesZero)
-        zero<true>(p, size);
+      {
+        ::bzero(p, size);
```

Comment:
We can only reach this if the `mmap` call above failed.  I don't think that `mmap` should be able to fail, so the error condition there should probably be `PAL::error` (which kills the process): if we can't `mmap` over some address space that we've previously allocated, then something is badly wrong and we're probably trying to `mmap` over some address space that we shouldn't be trying to `mmap` over so `bzero`ing it will make things even worse.

Reply:
PALPOSIX behaves in the same fashion. PALWindows calls `error` if `VirtualAlloc2FromApp` fails.

I completely agree with your comment, but I wanted to keep the inheritance from PALBSD and PALPOSIX, therefore I duplicated the behavior.

Also, Apple is now using undocumented errnos (e.g. mprotect can return EAGAIN on some production builds of macOS). So, I can't say for certain that a failed `MAP_FIXED` can occur only in fatal failure cases. Nor can I say that in the future `MADV_FREE_REUSABLE` will never cause `MAP_FIXED` to fail.

Reply:
I didn't want to deviate from PALPOSIX::notify_using's current behavior. `PALPOSIX::notify_using` will call `PALPOSIX::zero`. Currently if `mmap` fails, `bzero` is called on the pages that mmap failed to overwrite.

I wanted to maintain identical failure conditions as PALPOSIX.

Reply:
I feel that this is more of an issue of `PALPOSIX::notify_using`'s behavior. I too would like PALPOSIX and PALApple to abort the current process, like PALWindows. But I feel that belongs in a different pull request; one that will also update `PORTING.md` to include the invariant "failed `mmap` calls will cause the process to abort" or a something similar.

Reply:
That seems reasonable to move that to a different PR.